Improved REST error handling

jbertouch opened the pull request:
First pass at closing #15335. The new behavior for the `RestController#executeHandler` method is as follows:
- For a request to a **valid endpoint** with an **unsupported HTTP method** -> Return a _405 HTTP error_ including allowed methods for the endpoint in the **Allow** HTTP header. Refer to [HTTP/1.1 - 10.4.6 - 405 Method Not Allowed](https://tools.ietf.org/html/rfc2616#section-10.4.6).
- For an OPTIONS HTTP method request to a **valid endpoint** -> Return a _200 HTTP response_ including allowed methods for the endpoint in the **Allow** HTTP header. Refer to [HTTP/1.1 - 9.2 - Options](https://tools.ietf.org/html/rfc2616#section-9.2).

The tests extend `ESIntegTestCase` rather than the `ESSingleNodeTestCase`, as `http.enabled` is set to false in the latter class.

Let me know if you think the tests should be broader - they could be improved by pulling in a random selection of REST endpoints to test, rather than the existing hardcoded endpoints.

nik9000 commented:
Thanks for changing the title! I use those to triage what I review and the old title didn't help!

jbertouch commented:
No worries, I realized after I hit submit that the old title was pretty opaque, especially if you were scanning through the PR list.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -208,19 +210,43 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
-            }
-        }
-    }
+	void executeHandler(RestRequest request, RestChannel channel) throws Exception {
```
Sneaky indentation change snuck in.

nik9000 commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersIT.java:
```diff
@@ -0,0 +1,84 @@
+	 * <a href="https://tools.ietf.org/html/rfc2616#section-9.2">HTTP/1.1 - 9.2
+	 * - Options</a>).
+	 */
+	public void testValidEndpointOptionsResponseHTTPHeader() throws Exception {
```
s/Ehdpoint/Endpoint/

jbertouch replied:
A spelling error? I can't find 'Ehdpoint' in this file...

nik9000 commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersIT.java:
```diff
@@ -0,0 +1,84 @@
+	 */
+	public void testValidEndpointOptionsResponseHTTPHeader() throws Exception {
+		createIndex("test");
+		HttpResponse httpResponse = httpClient().method("OPTIONS").path("/test").execute();
```
s/"\/test"/"/obviously_invalid_we_will_never_put_anything_here_I_hope/ ?

nik9000 replied:
oh, this has to be valid, sorry, my mistake.

nik9000 commented:
I like it! @jasontedor filed the original bug and will probably want to look too!

jbertouch commented:
Thanks for the review @nik9000. Just realized my updates are tab indented - will change to space indents to match the ES code style. Sorry that snuck in, was working on something else with tab indents and forgot to change the style.

nik9000 commented:
> Sorry that snuck in,

It is cool! I suspect the build would warn you about that with `gradle precommit`.

jasontedor commented:
@jbertouch I gave it a first pass and it looks great.

Regarding

> The tests extend ESIntegTestCase rather than the ESSingleNodeTestCase, as http.enabled is set to false in the latter class.

you can override the `ESSingleNodeTestCase#nodeSettings` method to set `http.enabled` to `true`. Can you check if that would work?

jbertouch commented:
@jasontedor Thank you for the review.

I can't see a `ESSingleNodeTestCase#nodeSettings` method. Did you mean the `ESIntegTestCase#nodeSettings` method? If so, that's the one I overrode in the the `RestHTTPResponseHeadersIT` class.

jasontedor commented:
@jbertouch Nope, I'm referring to a [protected method on `ESSingleNodeTestCase`](https://github.com/elastic/elasticsearch/blob/3046055bd0f2bc3d334667b9d8cad2f6b7396cba/test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java#L169) that your test suite can override to return an instance of `Settings` that has set `http.enabled` to `true`.

jbertouch commented:
Ah, sorry I see it now. I hadn't merged in the latest commits from master in a while.

jbertouch commented:
@jasontedor Updated the test class to extend `ESSingleNodeTestCase` rather than `ESIntegTestCase`.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersTests.java:
```diff
@@ -0,0 +1,99 @@
+     * <a href="https://tools.ietf.org/html/rfc2616#section-10.4.6">HTTP/1.1 -
+     * 10.4.6 - 405 Method Not Allowed</a>).
+     */
+    public void testUnsupportedMethodResponseHTTPHeader() throws Exception {
```
Nit: `HTTP` -> `Http`

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersTests.java:
```diff
@@ -0,0 +1,99 @@
+     * <a href="https://tools.ietf.org/html/rfc2616#section-9.2">HTTP/1.1 - 9.2
+     * - Options</a>).
+     */
+    public void testValidEndpointOptionsResponseHTTPHeader() throws Exception {
```
Nit: `HTTP` -> `Http`

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersTests.java:
```diff
@@ -0,0 +1,99 @@
+ * methods on REST endpoints should respond with status code 405</a> for more
+ * information.
+ */
+public class RestHTTPResponseHeadersTests extends ESSingleNodeTestCase {
```
Nit: `HTTP` -> `Http`

jbertouch replied:
Old habits die hard!

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHTTPResponseHeadersTests.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    protected Settings nodeSettings() {
+        return Settings.builder().put(NetworkModule.HTTP_ENABLED.getKey(), true).put(super.nodeSettings()).build();
```
I prefer the string constant `http.enabled` here over the key. It helps catch inadvertent changes to the key. I've shared my thoughts on this [previously](https://github.com/elastic/elasticsearch/pull/17948#discussion_r60895487).

jbertouch replied:
And keep a deeper moat around the test code? Fair point.

jasontedor commented:
I like it and left some feedback. I like that we have the integration tests included here, but I wonder if it's possible to set up a unit test as well? The basic idea would be to mock a situation accepting a random collection of verbs and ensure that we do the right thing here on an unsupported verb.

jbertouch commented:
Thanks for the feedback @jasontedor. I will turn the test coverage up to 11, and add a unit test to complement the integration test.

jbertouch commented:
Hi all, this is ready for another review. I made the suggested tweaks to the integration test, and added randomizing unit test coverage for the new `RESTController` functionality.

jbertouch commented on core/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java:
```diff
@@ -0,0 +1,150 @@
+                return this;
+            }
+
+        }).initialize(invalidHttpMethodArray.get(0));
```
Alternatively, I'm happy to add an additional constructor to the `FakeRestRequest` class accepting a `RestRequest.Method` (and default the `RestRequest.Method` to `GET`).

jasontedor replied:
I wonder if this should be a random entry in `invalidHttpMethodArray`, just for thoroughness?

jasontedor commented:
> Hi all, this is ready for another review. I made the suggested tweaks to the integration test, and added randomizing unit test coverage for the new RESTController functionality.

Thanks. I'll give this a thorough review in the next few days, hopefully tomorrow.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java:
```diff
@@ -0,0 +1,150 @@
+         * the array, so we are guaranteed at least one invalid method to test.
+         */
+        validHttpMethodArray = validHttpMethodArray.subList(0, randomIntBetween(1, validHttpMethodArray.size() - 1));
+        assertTrue(validHttpMethodArray.size() > 0);
```
I think this should just be a plain Java `assert`.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java:
```diff
@@ -0,0 +1,150 @@
+         */
+        validHttpMethodArray = validHttpMethodArray.subList(0, randomIntBetween(1, validHttpMethodArray.size() - 1));
+        assertTrue(validHttpMethodArray.size() > 0);
+        assertTrue(validHttpMethodArray.size() < RestRequest.Method.values().length);
```
I think this should just be a plain Java `assert`.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestHttpResponseHeadersTests.java:
```diff
@@ -0,0 +1,150 @@
+         */
+        List<RestRequest.Method> invalidHttpMethodArray = new ArrayList<RestRequest.Method>(Arrays.asList(RestRequest.Method.values()));
+        invalidHttpMethodArray.removeAll(validHttpMethodArray);
+        assertTrue(invalidHttpMethodArray.size() > 0);
```
I think this should just be a plain Java `assert`.

jasontedor commented:
So this doesn't yet close #15335. For example:

```bash
$ curl -XGET localhost:9200/                                                                    
{
  "name" : "Mandrill",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "5.0.0",
    "build_hash" : "b08ee7f",
    "build_date" : "2016-05-11T00:26:27.003Z",
    "build_snapshot" : true,
    "lucene_version" : "6.0.0"
  },
  "tagline" : "You Know, for Search"
}
$ curl -I -XGET localhost:9200/_forcemerge                                                                                                                                                                                                                                                                 
HTTP/1.1 400 Bad Request
Content-Type: application/json; charset=UTF-8
Content-Length: 203
$ curl -XGET localhost:9200/_forcemerge?pretty=1
{
  "error" : {
    "root_cause" : [ {
      "type" : "illegal_argument_exception",
      "reason" : "No feature for name [_forcemerge]"
    } ],
    "type" : "illegal_argument_exception",
    "reason" : "No feature for name [_forcemerge]"
  },
  "status" : 400
}
```

Can anything be done here?

jbertouch commented:
Hi @jasontedor, thank you for the review. Nice catch on the `_forcemerge` edge case, wish I had picked that up.

Interesting that [`GetIndexRequest`](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexRequest.java#L80) appears to be handling the `_forcemerge` `GET` request. From a quick look at the code the error originates in [`RestGetIndicesAction#handleRequest`](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/rest/action/admin/indices/get/RestGetIndicesAction.java#L73). Looks like this error has manifested since [this commit](https://github.com/elastic/elasticsearch/commit/f709b7283f283aa2d3918cc1afda1adb81f26f47#diff-854597e8d375d3c783fa750672bec92b), which (correctly) removed the redundant GET method for `_forcemerge`.

The code implies that any `GET` request starting with an underscore will trigger the error you observed (eg. `curl -XGET http://localhost:9200/_alpha` or `curl -XGET http://localhost:9200/_alpha/beta`).

I will look into it. Would you prefer I open a separate PR with a fix for this specific issue, or include it with this PR?

jasontedor commented:
> Interesting that [`GetIndexRequest`](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexRequest.java#L80) appears to be handling the `_forcemerge` `GET` request.

Yup, that's why we are seeing the `No feature for name [_forcemerge]` message merge because it's acting like a request on an index that does not exist.

> The code implies that any `GET` request starting with an underscore will trigger the error you observed (eg. `curl -XGET http://localhost:9200/_alpha` or `curl -XGET http://localhost:9200/_alpha/beta`).

That's exactly right.

> I will look into it. Would you prefer I open a separate PR with a fix for this specific issue, or include it with this PR?

Let's try to get in this PR? 😄

jbertouch commented:
Sure, will do. I guess it does fall under the REST error umbrella ☔

jbertouch commented:
Hi @jasontedor, sorry about the delay in responding. A couple of quick thoughts on the greedy index matcher problem, aka the `_forcemerge` issue you identified.

- One possible solution is to modify the `PathTrie` class so that nodes are keyed to compiled regex Patterns, rather than Strings. This would allow more sophisticated path matching. The downside would be the performance hit from pointwise matching against all the child regex Patterns. Would need to benchmark this to be sure, the `RestController#executeHandler` is obviously invoked very frequently.
- An alternative would be to recognize when a handler has been incorrectly invoked (as with your `_forcemerge` example), and not execute it. This seems like a less robust solution to me. There could also be a performance impact from 'second-guessing' the matched handler before it's executed.

Interested in your thoughts on this.

jasontedor commented:
Sorry for the slow reply here, I just haven't come up with any wonderful ideas. But I wonder if it's as simple as if a handler is registered with an explicit name (e.g., `_forcemerge`) then no request matching `_forcemerge` can ever match any other handler except ones explicitly registered (i.e., let's prevent such requests from matching handlers with wildcards). Do you have any interest in seeing if that can be made to work (and doesn't break anything today)?

jbertouch commented:
Absolutely, I will look into it...

jasontedor commented:
> Absolutely, I will look into it...

Thanks @jbertouch. I was thinking about this again this morning and the more that I thought about it, the more that I think the approach I outlined above is a viable and good approach. I'll be curious to see what you come up with (the devil is always in the details) but this is great work so far that I'd really like to see us get in.

jbertouch commented:
Hi @jasontedor, I will find some time to work on this today or tomorrow, sorry about the delay.

elasticmachine commented:
Can one of the admins verify this patch?

jasontedor commented:
> I will find some time to work on this today or tomorrow, sorry about the delay.

I look forward to it, and no worries.

jbertouch commented:
Hi @jasontedor, this is ready for a review. Here's an outline of my approach to fix the explicit/wildcard path matching edge case issue:

**PathTrie**
- I added a parameter to `PathTrie#retrieve` to either include or ignore wildcard nodes when traversing the trie. This seemed like the path of least resistance to prioritizing explicit path matches (no pun intended).
- Wildcards in `PathTrie` leaf nodes are matched in both the include and ignore modes.
- Not matching wildcard leaf nodes when searching for explicit path matches triggered multiple REST integration test errors. For example, a simple GET request to `/_settings` returned a 405 error, as the `RestGetSettingsAction` handler registered to GET `/_settings/{name}` wasn't being assigned to the request. The `RestController` incorrectly assumed that the request was intended for `RestUpdateSettingsAction`, registered to PUT `/_settings`. Changing the handling of wildcard leaf nodes fixed this issue.

**RestController**
- `RestController#executeHandler` now prioritizes explicit handler matches. It falls back to wildcard path matched handlers if none are found. I refactored some of this logic into separate methods to make the flow a bit clearer.
- All the enhanced OPTIONS http method handling is still in there, plus unit/integration tests.

One other thing. 2 integration tests are failing with the same `illegal_state_exception` 500 error  - `:modules:transport-netty4:integTest` and `:qa:backwards-5.0:integTest`. I'm seeing the same exception as the one flagged in #19419, even after I merged the fix in PR #19432. I would be happy to check it out, but the error appears to be unrelated to these changes.

jbertouch commented:
I believe this PR will also fix #17853, which appears to be due to incorrect handler matching. Will confirm.

jasontedor commented:
ok to test

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -240,25 +239,156 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
+         * matches) for this request, for the set of HTTP methods.
+         */
+        final HashSet<RestRequest.Method> explicitPathValidMethodSet = getValidHandlerMethodSet(request, true);
+        if (explicitPathValidMethodSet.size() > 0 
```
I think that this block of code is identical to a block of code in `RestController#executeExplicitHandler`; can we collapse them into a method?

jasontedor commented on core/src/main/java/org/elasticsearch/common/path/PathTrie.java:
```diff
@@ -218,6 +254,10 @@ public T retrieve(String path, Map<String, String> params) {
             index = 1;
         }
-        return root.retrieve(strings, index, params);
+        if (ignoreWildcards == true) {
```
We can just write this as `if (ignoreWildcards)`; the `if (b == false)` thing is to avoid writing `if (!b)` because sometimes the `!` gets lost when reading, I guess.

jasontedor commented on core/src/main/java/org/elasticsearch/common/path/PathTrie.java:
```diff
@@ -175,6 +180,33 @@ public T retrieve(String[] path, int index, Map<String, String> params) {
         }
 
+        /**
+         * Retrive only explicitly mapped nodes (ie. wildcards are only allowed
```
Nit: `Retrive` -> `Retrieve`

jasontedor commented on test/framework/src/main/java/org/elasticsearch/test/rest/FakeRestRequest.java:
```diff
@@ -116,4 +116,4 @@ public FakeRestRequest build() {
     }
 
-}
+}
```
Can we keep the newline at the end of the file here?

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -328,4 +485,4 @@ public void process(RestRequest request, RestChannel channel, NodeClient client,
         }
     }
-}
+}
```
Can we keep the newline at the end of the file here?

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -181,10 +180,10 @@ public RestFilterChain filterChain(RestFilter executionFilter) {
     /**
      * @param request The current request. Must not be null.
-     * @return true iff the circuit breaker limit must be enforced for processing this request.
+     * @return true if the circuit breaker limit must be enforced for processing this request.
```
I think this should be kept as if and only if.

jasontedor commented:
@jbertouch Sorry for the long delay in reviewing this. This is great work, I think we are very close now. I left a few minor comments, just cosmetic fixes. However, you mentioned #17853 and I do not think that this one is addressed by this PR, it looks like requests to `POST /{index}/_settings` are still being routed to an index action, matching the `POST` handler for `/{index}/{type}` but ideally these would 405 since there is a `PUT` handler for `{index}/_settings`. Can anything be done here?

I would like to get this integrated this week. I can commit to setting aside time to iterate on reviewing and integrating this during this week.

jbertouch commented:
Hi @jasontedor, thank you for the review, I will get a revision back to you early this week. Will take a closer look at #17853 too.

jasontedor commented:
@jbertouch Any chance of getting this to the finish line?

jbertouch commented:
Hi @jasontedor, hope you're well. Sorry I haven't had time to devote to this recently - will review it over the weekend.

jbertouch commented:
@jasontedor I had some time to review issue #17853 on a flight between LA and Sydney. Here's a high-level outline of my updated solution:

**PathTrie**
- I added 4 trie matching modes for node retrieval: `EXPLICIT_NODES_ONLY`, `WILDCARD_ROOT_NODES_ALLOWED`, `WILDCARD_LEAF_NODES_ALLOWED` and `WILDCARD_NODES_ALLOWED`, and updated the recursive `PathTrie#retrieve` method to support them.
- These modes allow for more nuanced path matching, and can be used to selectively prevent wildcard nodes from greedily matching requests.
- Test cases for the updated matching functionality are in `PathTrieTests#testWildcardMatchingModes` and `PathTrieTests#testExplicitMatchingMode`.

**RestController**
- Requests now cascade through the 4 new `PathTrie.TrieMatchingMode` modes, from the most restrictive (ie. only explicit nodes are matched), to the least restrictive (ie. wildcards are allowed anywhere). Requests were formerly matched allowing wildcards anywhere in the path, which was the cause of the various errors we were seeing.
- If a matching mode fails to find a handler, or a handler for a different HTTP method using the same matching mode, then the request is passed to a more permissive matching mode.
- If a match on a different HTTP method is found, the request is passed to `RestController#handleUnsupportedHttpMethod`.
- If no matches are found, the request is passed to `RestController#handleBadRequest`.
- There is a test for the edge case identified in issue #17853 in `RestHttpResponseHeadersIT#testIndexSettingsPostRequest`.

jbertouch commented:
One other thing, I guess `RestController#canTripCircuitBreaker` should also be updated to use the new handler matching logic?

jbertouch commented:
@jasontedor, any chance of reviewing these changes? I can close the PR if it's redundant.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jasontedor commented:
@jbertouch I've asked @dakrone to take over reviewing this one; he'll work with you on getting this one across the finish line!

dakrone commented:
Hi @jbertouch! We would *really* love to get this PR in as soon as we can, so it can be released! Would you be adverse to me helping out by cherry-picking your commits to a new branch and bringing this up to date with the current master branch? I know Github has a feature where repo members can push to a pull request's branch to collaboratively develop, but I think this PR was opened prior to that feature, so I can't push and help resolve the conflicts without a new branch/pr.

jbertouch commented:
Hi @dakrone, that's great to hear - sorry about the delay in getting back to you. Please go ahead and move my commits to a new branch.

dakrone commented:
Thanks! I've opened #24437 to continue this so I'm going to close this one.
